# FIT1043 Introduction to Data Science

 ## <center> Assignment 2 </center>

## Yo Kogure

## 32134541

___

## 1. Introduction

This is a report summarizing my attempt on the Assignment 2 of FIT1043.
I will try to keep the report as detailed, clean, appropriately formatted and easy to read as possible.

When I am not citing a reference, that means that I am not copy-pasting or referring to other site. However I do refer to lecture notes and totorial exercises without stating it each time.

References are stated with APA format at the conclusion at the end.

This file mostly consists as follows:
1. Introduction
2. Supervised Learning
3. Classification
4. Kaggle attempts
5. Conclusion

#### I will now import necessary libraries, and read the file.

In [1]:
import numpy as np  # numpy helps us deal with numbers in the future
import pandas as pd  # pandas, as we talked in the assignment 1, it allows us deal with data and wrangle.
import matplotlib.pyplot as plt  # this library allows us to plot many different kind of graphs and visualize.

Later on as we need it, I will download some more libraries such as sklearn.model ( which I will be covering later). I only actually needed pandas here to load csv.

In [2]:
essay_data = pd.read_csv("FIT1043-Essay-Features.csv")  # loading csv file using pandas

In [3]:
essay_data.shape

(1332, 19)

The data has 1332 rows and 19 columns each.

In [4]:
essay_data.sample(10)

essayid  chars  words  commas  apostrophes  punctuations  \
677       469   3049    648      26           19             2   
442       431   3554    694       7            4             4   
121        79    650    133       3            5             0   
970      1640   1468    300       8            6             0   
300      1119   1774    375       4           19             1   
557       914   1979    390      13           14             0   
966       485   1095    204       5            6             0   
1064      529   1982    393       3            8             0   
1263      737   1598    341      27            3             0   
650      1391    952    206       0            2             0   

      avg_word_length  sentences  questions  avg_word_sentence         POS  \
677          4.705247         43          2          15.069767  643.659420   
442          5.121037         21          1          33.047619  691.325626   
121          4.887218          7          0          19.000000  131.984733   
970          4.893333         16          0          18.750000  299.328859   
300          4.730667         18          1          20.833333  373.329749   
557          5.074359         20          2          19.500000  389.329897   
966          5.367647         13          0          15.692308  198.979899   
1064         5.043257         24          3          16.375000  389.984655   
1263         4.686217          5          0          68.200000  337.994065   
650          4.621359         15          0          13.733333  203.990148   

      POS/total_words  prompt_words  prompt_words/total_words  synonym_words  \
677          0.993302           227                  0.350309            123   
442          0.996146           306                  0.440922            160   
121          0.992366            59                  0.443609             34   
970          0.997763           136                  0.453333             70   
300          0.995546           154                  0.410667            104   
557          0.998282           172                  0.441026             99   
966          0.975392            88                  0.431373             58   
1064         0.992327           187                  0.475827             98   
1263         0.991185           157                  0.460411             86   
650          0.990243            76                  0.368932             69   

      synonym_words/total_words  unstemmed  stemmed  score  
677                    0.189815        738      725      4  
442                    0.230548        744      719      3  
121                    0.255639        173      173      3  
970                    0.233333        371      357      3  
300                    0.277333        412      405      2  
557                    0.253846        466      445      3  
966                    0.284314        280      275      3  
1064                   0.249364        498      491      4  
1263                   0.252199        404      395      4  
650                    0.334951        243      238      1

I have randomly taken a sample of 5 rows.
To roughly give some description of the data I have read, I will not repeat what each of these columns indicate, because it is already done so in the specification sheet.

#### This sets of data contains information on 1332 different essays. It stores details of each essay regarding words and characters, such as total characters, average of each word length, number of synonyms, etc. There are 19 different columns so 19 different characteristics/features stored for this set of data.

#### Ultimately, I believe that our goal is to build a model to predict its "score" from 1-6, from sets of given features of each essay.
One that approximates y = f(x), where y is the score and x are all other features.

I will assume that data is already clean and wrangled to some point, because so far from what I have checked manually, they are, with no NaN values.

## 2. Supervised Learning

Supervised machine learning means, a part of machine learning feature that includes analyzing the data to find out a suitable function that maps a set of inputs to its corresponding outputs. For a set of x and y, its aim is to find a certain function that mostly accurately predicts value of y given a certain value of x.

One feature of "supervised learning" is that the algorithm uses labeled data, to improve its accuracy on the prediction of given data. It also efficiently use train datasets to analyze and test datasets to actually test the algorithm you have found out. 

What does labeled data mean? For example for the above data, its "raw" data contains thousands of datasets for each 19 different columns. They are just "unlabeled" numbers and even machines can not process until we label them and make it valuable for the machine learning. It extracts necessary fields, and adds meaningful label or tag to the raw data in order to perform supervised learning.

In order to apply supervised learning, datasets must be set to training datasets and testing datasets.
Training datasets are data that the algorithm takes in as an input to learn and find the function. Based on train data, an appropriate model is implemented. 
Test datasets are those that are used to check how accurate the function or model is, after algorithm finds out suitable model based on training datasets.

For example out of 100 sets of given data, if 100 are used for training, and use same 100 for testing, it doesn't allow us to test from neutral standpoint. Because that model is built by analyzing all 100 data in the first place, of course the test will return positive result. This won't be a fair accuracy test unless we divide them into training datasets and testing datasets.

---

In this context, features basically means property of your training data. It is like x values from which we will take in as input.

Label is like y values, in short. It is a set of output that we get after training with the model.

So for the label here is 'score', because we want to find and predict it as an output from 1-6 given set of other features.

I will seperate features and label using pd.iloc[] .

• essayid - a unique id to identify the essay <br>
• chars - number of characters in the essay, including spaces <br>
• words - number of words in the essay <br>
• commas - number of commas in the essay <br>
• apostrophes - number of apostrophes in the essay <br>
• punctuations - number of punctuations (other than commas, apostrophes, period,
questions marks in the essay <br>
• avg_word_length - the average length of the words in the essay <br>
• sentences - number of sentences in the essay, determined by the period (fullstops) <br>
• questions - number of questions in the essay, determined by the question marks <br>
• avg_word_sentence - the average number of words in a sentence in the essay <br>
• POS - total number of Part-of-Speech discovered <br>
• POS/total_words - fraction of the POS in the total number of words in the essay <br>
• prompt_words - words that are related to the essay topic <br>
• prompt_words/total_words - fraction of the prompt words in the total number of
words in the essay <br>
• synonym_words - words that are synonymous <br>
• synonym_words/total_words - fraction of the synonymous words in the total
number of words in the essay <br>
• unstemmed - number of words that were not stemmed in the essay <br>
• stemmed - number of words that were stemmed (cut to the based word) in the essay <br>
• score - the rating grade, ranging from 1 – 6 <br>
### Copied and pasted from specification sheet for reference.

 #### I am allowed to engineer or remove features as I deem appropriate,  so will be doing so at earlier stage.

I will be excluding essayid from features because it does not affect grade, they are just referencing numbers. For now only essayid.

In [5]:
del essay_data["essayid"]  # remove column

In [6]:
essay_data.head()

chars  words  commas  apostrophes  punctuations  avg_word_length  \
0   2153    426      14            6             0         5.053991   
1   1480    292       9            7             0         5.068493   
2   3964    849      19           26             1         4.669022   
3    988    210       8            7             0         4.704762   
4   3139    600      13            8             0         5.231667   

   sentences  questions  avg_word_sentence         POS  POS/total_words  \
0         16          0          26.625000  423.995272         0.995294   
1         11          0          26.545455  290.993103         0.996552   
2         49          2          17.326531  843.990544         0.994100   
3         12          0          17.500000  207.653784         0.988828   
4         24          1          25.000000  594.652150         0.991087   

   prompt_words  prompt_words/total_words  synonym_words  \
0           207                  0.485915            105   
1           148                  0.506849             77   
2           285                  0.335689            130   
3           112                  0.533333             62   
4           255                  0.425000            165   

   synonym_words/total_words  unstemmed  stemmed  score  
0                   0.246479        424      412      4  
1                   0.263699        356      345      4  
2                   0.153121        750      750      4  
3                   0.295238        217      209      3  
4                   0.275000        702      677      4

In [7]:
y = essay_data.iloc[:, -1].values
y  # y is the label, because it acts as an output, for "score"

array([4, 4, 4, ..., 3, 3, 4])

In [8]:
X = essay_data.iloc[:, :-1].values  # X contains all other features excluding score
X

array([[2.15300000e+03, 4.26000000e+02, 1.40000000e+01, ...,
        2.46478873e-01, 4.24000000e+02, 4.12000000e+02],
       [1.48000000e+03, 2.92000000e+02, 9.00000000e+00, ...,
        2.63698630e-01, 3.56000000e+02, 3.45000000e+02],
       [3.96400000e+03, 8.49000000e+02, 1.90000000e+01, ...,
        1.53121319e-01, 7.50000000e+02, 7.50000000e+02],
       ...,
       [1.81400000e+03, 3.63000000e+02, 5.00000000e+00, ...,
        2.94765840e-01, 4.27000000e+02, 4.15000000e+02],
       [1.42700000e+03, 2.87000000e+02, 5.00000000e+00, ...,
        2.89198606e-01, 3.23000000e+02, 3.12000000e+02],
       [2.80600000e+03, 5.42000000e+02, 2.40000000e+01, ...,
        2.85977860e-01, 5.96000000e+02, 5.75000000e+02]])

---

In [9]:
# I will import necessary library to split my data for training and testing
from sklearn.model_selection import train_test_split

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)  
# test_size = 0.25 means that 75% is used for training and 25% is used for testing
# random_state affects how to decide how they are seperated to training and testing. 
# Because it is =0, for this case they are seperated very randomly

In [11]:
len(X_train) , len(X_test), len(y_train), len(y_test)

(999, 333, 999, 333)

There! They are correctly seperated into training and testing datasets. The ratio can honestly be anything, 80:20, 70:30 or 90:10 depending on the situaition, but I would like to take 75:25 as a balanced middle value.

In [12]:
X_train[:3], X_test[:3], y_train[:3], y_test[:3]

(array([[2.34500000e+03, 4.90000000e+02, 9.00000000e+00, 1.10000000e+01,
         0.00000000e+00, 4.78571429e+00, 1.70000000e+01, 0.00000000e+00,
         2.88235294e+01, 4.86991786e+02, 9.93860789e-01, 2.04000000e+02,
         4.16326531e-01, 1.27000000e+02, 2.59183673e-01, 5.07000000e+02,
         4.95000000e+02],
        [3.01000000e+03, 6.21000000e+02, 4.80000000e+01, 7.00000000e+00,
         0.00000000e+00, 4.84702093e+00, 2.80000000e+01, 1.00000000e+00,
         2.21785714e+01, 6.16987076e+02, 9.93537964e-01, 3.15000000e+02,
         5.07246377e-01, 1.44000000e+02, 2.31884058e-01, 6.72000000e+02,
         6.61000000e+02],
        [2.09900000e+03, 4.20000000e+02, 1.10000000e+01, 5.00000000e+00,
         0.00000000e+00, 4.99761905e+00, 2.00000000e+01, 0.00000000e+00,
         2.10000000e+01, 4.15320513e+02, 9.88858364e-01, 2.52000000e+02,
         6.00000000e-01, 1.25000000e+02, 2.97619048e-01, 4.47000000e+02,
         4.42000000e+02]]),
 array([[1.87000000e+03, 3.71000000e+02, 2.1

#### Looks like it succeeded. Moving on to the next part of the assignmnet.

---

## 3. Classification

Binary classification is when the given dataset is categorized into 2 distinct classes in the supervised learning.

For multi-class classification, as its name suggests, they are not categorized into 2, but more than 2, multiple classes.

This concept is fairly simple, but to give an example, if we try to implement a model to predict whether data showing whether students failed FIT1043 from given data, it is considered as binary classification, because only 2 distinct choices in the label.

But a model to predict which grade that student took, is considered as multi-class classification, because there are multiple options to classify them like HD, D, N and F.

---

This part of question requires knowledge of normalising of data. From that fact, I can predict that later on I will be using Support Vector Machine/Regression to deal with several multiple features more than 1(which I kind of already know). And I can also tell that some values of the input are either too large or too small. I will explain why I thought so.

<br>

The reason for normalising the data, is because some values for columns are just too large or too small. After normalisation, different values that were measured under different scales are set to a common scale, with 0 being the value in the common scale for the average value. Many classifiers that are in python libraries calculate the distance between two points by using Euclidean distance.

In [13]:
# For example,
small_test = essay_data.filter(items=['chars', 'commas', 'questions','score'])
small_test.sample(5)

chars  commas  questions  score
734   2569      33          0      5
119   2221      11          1      4
280   2108      14          0      4
225   2020       5          0      4
705    173       0          0      1

Note that we are unaware of weighting of importance of each features on the score. However, if we were to execute classification by finding distance between two points, (imagine drawing lines from each point to point and using pythagoras theorem to find its distance)values in chars will be too dominant. We want to consider commas, questions and of course other features fairly. They can't be too large or too small.

In such case features do not contribute equally to our output, hence we have a need to re-scale them by using normalisation.

In [14]:
from sklearn.preprocessing import StandardScaler
# Import necessary library for standardisation 

In [15]:
# I will apply scaling and transform its data in training datasets only.

In [16]:
sc = StandardScaler()  # StandardScaler() from sklearn.preprocessing is the function that I will be using
X_train = sc.fit_transform(X_train)  # Train cases standardised, and transformed
X_test = sc.transform(X_test)  
# The mean and standard deviation value of train case is stored in sc, and applied to test cases now

In [17]:
X_train[:2], X_test[:2]

(array([[ 0.28684034,  0.38716127, -0.52496032,  0.44571172, -0.41870753,
         -0.70522038, -0.23728237, -0.65162306,  0.40195483,  0.39407748,
          0.52980282,  0.07996837, -1.03601228,  0.38954264, -0.12344141,
          0.24983146,  0.26442177],
        [ 1.03456971,  1.13092488,  2.98026314, -0.20693175, -0.41870753,
         -0.4245677 ,  0.97890044, -0.12065125, -0.15102251,  1.13589011,
          0.48749963,  1.40547103,  0.76146475,  0.76766351, -0.83510202,
          1.2666706 ,  1.3115888 ]]),
 array([[-0.24725206, -0.28847132,  0.55356998, -0.85957522, -0.41870753,
          0.46083592,  0.315528  , -0.65162306, -0.59331856, -0.28880026,
          0.03598027, -0.06332921,  0.96457761, -0.34445669, -0.27501425,
         -0.38492267, -0.40425116],
        [-0.62167894, -0.72564535,  0.01430483, -0.04377088, -0.41870753,
          1.3189993 , -0.56896859,  0.41032056, -0.24910037, -0.71296754,
          0.88551956, -0.87534886, -0.92842096, -1.05621361, -1.38253326,
  

------

I will use Support Vector Machine/Regression algorithm to build the model.

Support vector machines, SVM are a set of supervised learning methods used for classification as well as regression algorithms. One of the benefits of SVM is that it can create the "best decision boundary" to separate two or more classes so that we can correctly place data points categorized into the correct class. 

One of its advantage is that it is effective is higher-dimentional spaces, that is, our features containing more than 10 different characteristics. Also I have learnt that it works relatively well in cases that there is a clear margin of seperation between classes. Since our label score are distinct integers ranging from 1-6, it also satisfies for our case.

(The question asked in relation to Linear Regression): <br>
Not only classification, but also for regression Support Vector Machine is useful. One of its advantage compared to usual Linear Regression is that SVM/SVR allows us to define for ourselves how much error is acceptable in our model and will find an appropriate line that fit the data.

<br>
*Reference:* <br>
https://dataaspirant.com/svm-kernels/ <br>
https://dhirajkumarblog.medium.com/top-4-advantages-and-disadvantages-of-support-vector-machine-or-svm-a3c06a2b107
https://towardsdatascience.com/an-introduction-to-support-vector-regression-svr-a3ebc1672c2

<br>
There is something called kernel in Support Vector Machine/ Support Vector Regression.
In my own words, you can consider kernel as a function in SVM that maps features to labels in supervised learning. Kernel machines are a class of algorithms for pattern analysis, and algorithms use a set of mathematical functions that are defined as the kernel.
There are different kinds of kernels, and they include polynomial kernel, gaussian kernel, etc. <br>

<br>
*Reference:* <br>
https://data-flair.training/blogs/svm-kernel-functions/ <br>
https://en.wikipedia.org/wiki/Kernel_method <br>

---

I have referred to: https://scikit-learn.org/stable/modules/svm.html for implementation of the model

In [18]:
len(X_train) , len(X_test), len(y_train), len(y_test)

(999, 333, 999, 333)

In [19]:
from sklearn import svm  # import support vector machine

Parameters for the built-in function: <br>
- An array X of shape = (n_samples, n_classes), where n_samples is number of samples and n_classes is number of features. This is basically X_train that we have defined earlier.
- An array y of class variable of shape (n_samples)

In [20]:
X_train.shape, y_train.shape

((999, 17), (999,))

In [21]:
clf = svm.SVC()  # Create an object clf of class SVC, basically we will now start applying the algorithm
clf.fit(X_train, y_train)  # It has been fitted

SVC()

### Model has been successfully created and its information saved.

## I will now conduct a prediction for the 'score', using testing dataset

In [22]:
y_pred = clf.predict(X_test)  # apply the prediction on X_test dataset and name it as y_pred
y_pred

array([4, 3, 3, 4, 3, 4, 4, 4, 2, 4, 4, 3, 4, 3, 3, 4, 3, 4, 4, 4, 3, 4,
       3, 3, 3, 4, 4, 4, 3, 4, 4, 3, 3, 4, 4, 3, 4, 4, 3, 3, 4, 2, 3, 3,
       3, 3, 3, 3, 2, 3, 4, 3, 4, 4, 4, 4, 4, 3, 3, 4, 3, 4, 3, 3, 4, 4,
       4, 4, 4, 4, 3, 3, 4, 5, 4, 3, 3, 4, 4, 3, 4, 3, 4, 3, 3, 4, 4, 3,
       4, 4, 3, 4, 3, 3, 4, 4, 3, 3, 4, 2, 4, 4, 3, 4, 4, 3, 4, 4, 3, 4,
       2, 3, 3, 3, 3, 4, 3, 4, 3, 3, 3, 3, 4, 4, 4, 3, 3, 4, 4, 4, 4, 4,
       4, 3, 4, 3, 3, 4, 4, 3, 4, 4, 4, 3, 3, 3, 3, 3, 4, 4, 4, 3, 4, 4,
       4, 3, 3, 3, 4, 3, 4, 4, 3, 3, 4, 4, 4, 4, 3, 3, 3, 3, 3, 2, 4, 3,
       4, 4, 4, 2, 3, 4, 4, 3, 3, 4, 3, 3, 3, 3, 3, 4, 3, 3, 2, 3, 4, 4,
       3, 4, 3, 4, 4, 3, 4, 4, 3, 3, 3, 3, 4, 4, 3, 3, 3, 3, 4, 4, 3, 3,
       4, 4, 4, 4, 4, 3, 3, 4, 3, 3, 4, 3, 4, 3, 3, 4, 4, 4, 4, 4, 4, 4,
       3, 3, 3, 4, 4, 3, 4, 4, 3, 4, 3, 4, 4, 4, 3, 2, 2, 3, 3, 3, 4, 3,
       3, 3, 4, 4, 4, 3, 3, 4, 3, 4, 4, 3, 3, 3, 4, 3, 4, 4, 2, 3, 4, 3,
       3, 3, 4, 4, 4, 4, 3, 4, 3, 3, 4, 4, 4, 3, 3,

In [23]:
from sklearn.metrics import confusion_matrix  # Import confusion matrix

In [24]:
cm = confusion_matrix(y_test, y_pred)
cm

array([[  0,   2,   0,   0,   0,   0],
       [  0,   8,  13,   2,   0,   0],
       [  0,   1, 103,  43,   0,   0],
       [  0,   0,  34, 109,   1,   0],
       [  0,   0,   0,  16,   0,   0],
       [  0,   0,   0,   1,   0,   0]])

This is our confusion matrix. It is a 6x6 matrix because our label, score has discrete choices of 6: 1, 2, 3, 4, 5, 6. If I were to explain some of the elements here, notice 103 at cm[2, 2]. This means that there are 103 data in testing dataset that we predicted is 3, and actually turned out to have a score of 3. 43 at cm[2, 3] shows that there are 43 test cases where we predicted them to have a score of 4, but actually had 3. You can look at diagnoals in the N*N matrix to see which one you have accurately predicted.

Quadratic Weighted Kappa(QWK) measures an agreement between two ratings. In our case two ratings mean our predicted scores and the actual/know scores. Agreement means that I have predicted a certain classification correctly, while disagreement means I have failed to do so. This metric typically varies from 0 (random agreement between raters) to 1 (complete agreement between raters). In cases such that there is less agreement between the raters than expected, the metric may go below 0. 


<br>
Reference:<br>
https://www.kaggle.com/aroraaman/quadratic-kappa-metric-explained-in-5-simple-steps

In [25]:
# To find Quadratic Weighted Kappa, I will use cohen_kappa_score from sklearn.metrics library

In [26]:
from sklearn.metrics import cohen_kappa_score

In [27]:
# sklearn.metrics.cohen_kappa_score(y1, y2, *, labels=None, weights=None, sample_weight=None)
kappa_value = cohen_kappa_score(y_test, y_pred, labels=None, weights='quadratic', sample_weight=None) 
# weights = quadratic because this is quadratic weighter kappa
kappa_value
# Refered to: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html

0.5735877574154544

### I have successfully obtained the QWK Score!

---

## 4. Kaggle Submission

I will read the file and use the model I built earlier to predict the 'score'

In [28]:
kaggle = pd.read_csv('FIT1043-Essay-Features-Submission.csv')

In [29]:
kaggle.head()

essayid  chars  words  commas  apostrophes  punctuations  avg_word_length  \
0     1623   4332    900      28           13             0         4.813333   
1     1143   1465    280      11            3             1         5.232143   
2      660   1696    325      17            2             0         5.218462   
3     1596   2640    555      20           17             0         4.756757   
4      846   2844    596      33            4             1         4.771812   

   sentences  questions  avg_word_sentence         POS  POS/total_words  \
0         39          1          23.076923  893.988852         0.993321   
1         14          3          20.000000  278.321343         0.994005   
2         19          1          17.105263  321.316770         0.988667   
3         28          0          19.821429  551.989150         0.994575   
4         24          9          24.833333  593.658810         0.996072   

   prompt_words  prompt_words/total_words  synonym_words  \
0           392                  0.435556            196   
1           131                  0.467857             51   
2           178                  0.547692             92   
3           228                  0.410811            107   
4           279                  0.468121            138   

   synonym_words/total_words  unstemmed  stemmed  
0                   0.217778        750      750  
1                   0.182143        339      316  
2                   0.283077        352      337  
3                   0.192793        632      605  
4                   0.231544        626      607

In [30]:
kaggle.shape  # The kaggle data seems good

(199, 18)

In [31]:
essayid = kaggle.iloc[:,0].values  # extract essayid, because I can't use it to apply the model I built earlier
essayid

array([1623, 1143,  660, 1596,  846,  868,  145,  500,  652,  214, 1229,
       1588, 1435,  115,  309,   70, 1068,   42,   47,  100,  165,  479,
        977, 1292,  764,   31,   25, 1438, 1216,  712,  368, 1180, 1514,
       1172, 1139,  835,  790,  893,  579, 1517, 1749,  402, 1570, 1049,
        925, 1140,  138,  671,  541,  889,  845, 1217, 1148,  197, 1087,
        612,  757, 1508,  720, 1685,  665,  738,  994, 1235,  576,  373,
        851, 1323, 1648,  810,  569,   89, 1552,  352,  472, 1647,  740,
        844, 1116,  921,  330,  658,  185, 1313,  849,   18,  662,  331,
        279,  124,  182,  811,  396,  708,  523,   36,  956,  785,  528,
        311,  876, 1198,  721, 1400,   72,  486, 1056,  906, 1048,  152,
        228, 1371,   37, 1754, 1769,  186, 1332,  928, 1446,  532,  980,
         91, 1224,  506, 1357, 1374, 1078,  983,  759, 1650,  635,  816,
       1488, 1600, 1268, 1285, 1424,  156, 1582,  886, 1105,  258, 1748,
       1007, 1617,  916, 1271, 1124,  631,  345,  6

In [32]:
k_test = kaggle.iloc[:, 1:].values  # remove essayid column from it

In [33]:
k_test = sc.transform(k_test)  # Standardise it

In [34]:
kaggle_pred = clf.predict(k_test)  # Predict by applying the model that I found earlier
kaggle_pred

array([4, 3, 3, 4, 4, 4, 3, 3, 3, 3, 3, 4, 4, 3, 4, 4, 4, 4, 4, 3, 3, 3,
       4, 4, 4, 4, 4, 4, 4, 3, 3, 4, 3, 3, 4, 3, 5, 4, 4, 3, 3, 4, 3, 3,
       2, 3, 3, 4, 4, 3, 3, 4, 4, 4, 3, 4, 4, 4, 4, 4, 3, 4, 3, 4, 3, 4,
       4, 4, 3, 4, 3, 4, 4, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 3, 4, 3, 4, 4,
       2, 3, 3, 3, 4, 3, 4, 4, 4, 4, 4, 3, 3, 4, 3, 3, 3, 4, 3, 3, 4, 4,
       4, 3, 4, 4, 4, 3, 3, 4, 2, 3, 4, 4, 4, 3, 3, 4, 3, 4, 4, 3, 4, 4,
       2, 3, 4, 4, 3, 4, 2, 4, 4, 4, 4, 3, 3, 4, 4, 3, 4, 4, 4, 3, 4, 3,
       4, 3, 3, 3, 4, 4, 4, 3, 2, 3, 4, 4, 3, 3, 2, 4, 3, 4, 4, 3, 4, 4,
       4, 3, 3, 4, 4, 2, 4, 4, 4, 4, 3, 3, 3, 3, 3, 4, 3, 4, 3, 4, 4, 3,
       3])

In [35]:
array_joined = np.vstack((essayid, kaggle_pred))    # merge 2 arrays together, but vstack. not concatenate.
array_joined  

array([[1623, 1143,  660, 1596,  846,  868,  145,  500,  652,  214, 1229,
        1588, 1435,  115,  309,   70, 1068,   42,   47,  100,  165,  479,
         977, 1292,  764,   31,   25, 1438, 1216,  712,  368, 1180, 1514,
        1172, 1139,  835,  790,  893,  579, 1517, 1749,  402, 1570, 1049,
         925, 1140,  138,  671,  541,  889,  845, 1217, 1148,  197, 1087,
         612,  757, 1508,  720, 1685,  665,  738,  994, 1235,  576,  373,
         851, 1323, 1648,  810,  569,   89, 1552,  352,  472, 1647,  740,
         844, 1116,  921,  330,  658,  185, 1313,  849,   18,  662,  331,
         279,  124,  182,  811,  396,  708,  523,   36,  956,  785,  528,
         311,  876, 1198,  721, 1400,   72,  486, 1056,  906, 1048,  152,
         228, 1371,   37, 1754, 1769,  186, 1332,  928, 1446,  532,  980,
          91, 1224,  506, 1357, 1374, 1078,  983,  759, 1650,  635,  816,
        1488, 1600, 1268, 1285, 1424,  156, 1582,  886, 1105,  258, 1748,
        1007, 1617,  916, 1271, 1124, 

In [36]:
final_submission = pd.DataFrame(array_joined)     # convert to dataframe
final_submission = final_submission.transpose()   # flip the row and column.

In [37]:
final_submission.columns = ['essayid', 'score']
final_submission.head()

essayid  score
0     1623      4
1     1143      3
2      660      3
3     1596      4
4      846      4

### Looks great!

In [38]:
final_submission.to_csv('32134541-YoKogure-v1.csv', index=False)  # Exported to csv file
# Refered to: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html

## Submission complete. QWK of 0.62886 on the first try, I will try a little more.

## 5. Conclusion

This concludes the end of my assignment 2. Before I tackle this assignment, I thought it was really hard because it was scary looking in the specification sheet. However, after looking through lectures, tutorial and actually trying these problems, I found out that I don't really have to be scared of this topic. It was more interesting and fun than I thought. 
<br>

It was mindblowing that Python I use everyday contained the machine learning library, and could deal with this much data and analyze in very intuitive way. This gave me inspirations, and I look forward to upcoming R language. I had a fun time. Thank you for reading.


## 32134541 Yo Kogure

### References:

#### 

- Saumya Awasthi. (2020). SEVEN MOST POPULAR SVM KERNELS. Retrieved from https://dataaspirant.com/svm-kernels/ <br>
- Dhiraj K. (2019). Top 4 advantages and disadvantages of Support Vector Machine or SVM. Retrieved from https://dhirajkumarblog.medium.com/top-4-advantages-and-disadvantages-of-support-vector-machine-or-svm-a3c06a2b107 <br>
- Tom Sharp. (2020). An Introduction to Support Vector Regression (SVR). Retrieved from https://towardsdatascience.com/an-introduction-to-support-vector-regression-svr-a3ebc1672c2 <br>
- Kernel Functions-Introduction to SVM Kernel & Examples. Retrieved from https://data-flair.training/blogs/svm-kernel-functions/ <br>
- Kernel_method. Retrieved from https://en.wikipedia.org/wiki/Kernel_method <br>
- 1.4. Support Vector Machines. Retrieved from https://scikit-learn.org/stable/modules/svm.html <br>
- sklearn.metrics.cohen_kappa_score. Retrieved from https://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html <br>
- pandas.DataFrame.to_csv. Retrieved from https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html <br>